transactionality in POH: rename with_active_record_rescue to with_active_record_transaction_and_rescue, have it wrap its yield in a transaction

jmartin-sul opened the pull request:
since that method is already wrapping high-level groups of DB operations which go together to form what should be a single atomic update

test transactions by forcing one part of the database updating to fail, after the prior step has succeeded, and check that the prior step was rolled back.

* the first commit is just a bit of log message clean up that i couldn't resist
* the second adds transactionality to the initial database seeding of endpoints etc
* the third is a bit of refactoring that sets up
* the fourth, which just adds transactionality to the method that already wraps high-level sets of DB operations

my plan is to put up a follow-on PR that maybe fleshes out the tests for this a bit more (#324), as well as adding a `db/README.md` (#281) which gives a bit of basic guidance about when and how to wrap sets of operations in transactions, etc.

closes #203

jmartin-sul commented on app/services/preserved_object_handler.rb:
```diff
@@ -162,24 +153,37 @@ def create_db_objects(status, validated=false)
+    results.concat(upd_results)
+  end
+
+  def confirm_version_on_db_objects
```
i am not very happy with this name, since it's so similar to `confirm_version_on_db_object`, which it calls.  i'm totally open to suggestions for better names.

ndushay replied:
how about `confirm_version_in_db` or `confirm_version_in_catalog` ?

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -91,14 +91,8 @@ def confirm_version
-        results.concat(confirm_version_on_db_object(pc_db_object, :version))
-      end
-      results.concat(confirm_results)
+      Rails.logger.debug "confirm_version #{druid} called"
```
why remove the specificity of the debug logging messages when they are still in the same different conditions?

ndushay commented on app/services/preserved_object_handler.rb:
```diff
@@ -110,14 +104,11 @@ def update_version
-          results.concat update_online_version
-        elsif endpoint.endpoint_type.endpoint_class == 'archive'
-        end
+      Rails.logger.debug "update_version #{druid} called"
```
why remove the specificity of the debug logging messages when they are still in the same different conditions?

jmartin-sul replied:
at the time of the message, it's not clear that the object actually exists, since we haven't checked for that.  though i believe we were checking for existence when that log message was initially added.

ndushay commented on spec/services/preserved_object_handler_spec.rb:
```diff
@@ -335,4 +335,42 @@
+      allow(Status).to receive(:default_status).and_return(nil)
+      po_handler.create
+      expect(PreservedObject.where(druid: druid)).not_to exist
+    end
```
should this "create" test go in the preserved_object_handler_create_spec  file?

ndushay commented on spec/services/preserved_object_handler_spec.rb:
```diff
@@ -335,4 +335,42 @@
     it_behaves_like 'PreservedCopy does not exist', :confirm_version
   end
+
+  describe '#with_active_record_transaction_and_rescue' do
```
should there be a test for each of:

- create
- create_after_validation
- update_version
- update_version_after_validation (doesn't exist yet)
- confirm_version

proving that they roll back when they are supposed to?  Or maybe proving that they use the method?

jmartin-sul replied:
i would actually like to do that, though i was having a bit of trouble coming up with a test for `update_version` that didn't feel too contrived.  but certainly possible.  at worst, we could force the pres obj update call to blow up, and confirm that the pres copy changes got rolled back.

i was also thinking it might be nice to test the private method specifically/explicitly, but maybe that's overkill?  see #324.

if we did add tests for the methods you suggested, would you prefer to do that as part of this PR, or as part of a follow on ticket (e.g. slightly change or expand the scope of #324)?

ndushay replied:
I think you could either expand the scope of #324 or make a new ticket.

jmartin-sul replied:
expanded scope of #324

jmartin-sul commented:
@ndushay, i think i've addressed all of your feedback (implemented the suggestions for method naming and test re-location, ticketed the suggested test improvements).  ready for another round of evaluation.
